Simplified/AxisService/NYPLAxisContentDownloader.swift

```diff
@@ -0,0 +1,163 @@
+    self.tasks = ThreadSafeDictionaryContainer(queue: queue)
+    self.handlers = ThreadSafeDictionaryContainer(queue: queue)
+    self.numberOfAttempts = ThreadSafeDictionaryContainer(queue: queue)
+    self.errored = ThreadSafeValueContainer(value: false, queue: queue)
+    self.networkExecutor = networkExecuting
+  }
+  
+  func downloadItem(from url: URL,
+                    _ completion: @escaping (Result<Data, Error>) -> Void) {
+    
+    // We dont want to download anything if it's already being downloaded or if
+    // we experienced an error downloading something before.
+    guard !errored.value, tasks[url] == nil, handlers[url] == nil else {
+      return
+    }
+    
+    numberOfAttempts[url] = (self.numberOfAttempts[url] ?? 0) + 1
+    handlers[url] = completion
+    tasks[url] = self.getTask(for: url)
+  }
+  
+  private func getTask(for url: URL) -> URLSessionDataTask {
+    
+    let request = URLRequest(url: url,
+                             cachePolicy: .reloadIgnoringLocalCacheData,
+                             timeoutInterval: networkExecutor.requestTimeout)
+    
+    let task = networkExecutor.GET(request) { [weak self] (result) in
+      guard let self = self else { return }
+      switch result {
+      case .success(let data):
+        self.handlers[url]?(.success(data))
+        self.tasks[url] = nil
+        self.handlers[url] = nil
+      case .failure(let error):
+        // In poor network conditions, the download might fail on first or
+        // second attempt. We retry 2 more times to make sure we don't
+        // prematurely stop the download process.
+        if
+          let attempts = self.numberOfAttempts[url],
+          attempts < 3,
+          let handler = self.handlers[url] {
+          print("NYPLAxisContentDownloader downloading item from \(url.absoluteString) after \(attempts) failed attempts")
+          self.tasks[url] = nil
+          self.handlers[url] = nil
+          self.downloadItem(from: url, handler)
+          return
+        }
+        
+        // Here we perform short-circuiting so that we don't download anything
+        // else after receiving a failure since a book with a missing file is
+        // unusable.
+        self.handlers[url]?(.failure(error))
+        self.tasks.forEach {
+          $1.cancel()
+        }
+        self.tasks.removeAll()
+        self.handlers.removeAll()
+        self.errored.value = true
+      }
+    }
+    
+    return task
+  }
+  
+}
+
+private class ThreadSafeContainer {
+  fileprivate let queue: DispatchQueue
+  
+  init(queue: DispatchQueue) {
+    self.queue = queue
+  }
+  
+}
+
+private class ThreadSafeDictionaryContainer<S: Hashable,T>: ThreadSafeContainer {
+  
+  private var dictionary: [S:T]
+  
+  init(dictionary:[S:T] = [:], queue: DispatchQueue) {
+    self.dictionary = dictionary
+    super.init(queue: queue)
+  }
+  
+  subscript(key: S) -> T? {
+    get {
+      return queue.sync {
```

Comment:
I think these methods have retain cycles (sorry I didn't notice this sooner): self > queue property > closure > self

Reply:
@ettore I originally considered using weak self but didn't since these blocks are short lived and `self` would be released as soon as the value is returned/modified thus breaking the retain cycle. However, not having retain cycles is still better than short lived retain cycles. Fixed.